nestjs/src/listeners/course.listener.ts

```diff
@@ -0,0 +1,118 @@
+import { S3 } from '@aws-sdk/client-s3';
+import { Course } from '@entities/course';
+import { HttpService } from '@nestjs/axios';
+import { Injectable, Logger } from '@nestjs/common';
+import { Cron, CronExpression } from '@nestjs/schedule';
+import { InjectRepository } from '@nestjs/typeorm';
+import { ExportCourseDto } from 'src/courses/dto';
+import { FindOptionsWhere, MoreThanOrEqual, Repository } from 'typeorm';
+import { ConfigService } from '../config';
+
+@Injectable()
+export class CourseListener {
+  private s3: S3;
+  private logger = new Logger(CourseListener.name);
+
+  private readonly objectKey = 'app/courses.json';
+  private readonly bucket;
+
+  constructor(
+    @InjectRepository(Course)
+    private readonly repository: Repository<Course>,
+    private readonly httpService: HttpService,
+    readonly configService: ConfigService,
+  ) {
+    this.s3 = new S3(this.configService.awsClient);
+    this.bucket = this.configService.buckets.cdn;
+  }
+
+  @Cron(CronExpression.EVERY_30_MINUTES)
+  async handleCoursesChange() {
+    // check if courses are changed since last export
+    const changed = await this.areCoursesChanged();
+    if (!changed) {
+      return;
+    }
+
+    // export courses to cdn
+    await this.exportCoursesToCdn();
+
+    // trigger rs site update
+    this.triggerSiteUpdate();
+  }
+
+  /**
+   * Check if courses are changed since last export
+   * @returns true if courses are changed, false otherwise
+   */
+  private async areCoursesChanged() {
+    const response = await this.s3.headObject({ Bucket: this.bucket, Key: this.objectKey });
+    const lastModifiedCourse = await this.repository.findOne({
+      where: this.getCoursesWhereFilter(),
+      order: { updatedDate: 'DESC' },
+    });
+    return (response.LastModified?.getTime() ?? 0) < (lastModifiedCourse?.updatedDate.getTime() ?? 0);
```

Comment:
In general this is working solution as long as we have one source of updates for s3 object, but IMO it's not very reliable (e.g. object was manually re-uploaded or we'll have 2 ways to update this this object in future – and CDN conent can be mistakenly considered "newer" than courses in DB).  
As an alternative I can suggest following flow:  
1. Get courses from DB and CDN
2. Map courses from DB to `ExportCourseDto`
3. Calculate hash for both sources:  
  a) if hashes are equal – skip CDN update
  b) if hashes differ – update CDN  
This solution is slightly slower, but taking into account that amount of data is small and most likely won't grow significantly in the nearest future + this is background CRON job I think it's acceptable.

Reply:
@AlreadyBored thanks. I can't say I agree with your arguments but i liked the idea :) It makes code simplier. Also I don't do hash - just string comparison. Updated